tests/kafkatest/tests/streams/streams_cooperative_rebalance_upgrade_test.py

```diff
@@ -0,0 +1,191 @@
+        assert len(overlapping_tasks) == int(0), \
+            "Final task assignments are not unique %s %s %s" % (processor1.active_tasks, processor2.active_tasks, processor3.active_tasks)
+
+        other_overlapping_tasks = processor2.active_tasks.intersection(processor3.active_tasks)
+        assert len(other_overlapping_tasks) == int(0), \
+            "Final task assignments are not unique %s %s" % (processor2.active_tasks, processor3.active_tasks)
+
+        # test done close all down
+        stop_processors(processors, self.second_bounce_phase + self.stopped_message)
+
+        self.producer.stop()
+        self.kafka.stop()
+        self.zookeeper.stop()
+
+    def maybe_upgrade_rolling_bounce_and_verify(self,
+                                                processors,
+                                                previous_phase,
+                                                current_phase,
+                                                upgrade_from_version=None):
+        for processor in processors:
+            # stop the processor in prep for setting "update.from" or removing "update.from"
+            verify_stopped(processor, previous_phase + self.stopped_message)
+            # upgrade to version with cooperative rebalance
+            processor.set_version("")
+            processor.set_upgrade_phase(current_phase)
+
+            if upgrade_from_version is not None:
+                # need to remove minor version numbers for check of valid upgrade from numbers
+                upgrade_version = upgrade_from_version[:upgrade_from_version.rfind('.')]
+                rebalance_mode_msg = self.cooperative_turned_off_msg % upgrade_version
+            else:
+                upgrade_version = None
+                rebalance_mode_msg = self.cooperative_enabled_msg
+
+            self.set_props(processor, upgrade_version)
+            node = processor.node
+            with node.account.monitor_log(processor.STDOUT_FILE) as stdout_monitor:
+                with node.account.monitor_log(processor.LOG_FILE) as log_monitor:
+                    processor.start()
+                    # verify correct rebalance mode either turned off for upgrade or enabled after upgrade
+                    log_monitor.wait_until(rebalance_mode_msg,
+                                           timeout_sec=60,
+                                           err_msg="Never saw '%s' message " % rebalance_mode_msg + str(processor.node.account))
+
+                # verify rebalanced into a running state
+                rebalance_msg = current_phase + self.running_state_msg
+                stdout_monitor.wait_until(rebalance_msg,
+                                          timeout_sec=60,
+                                          err_msg="Never saw '%s' message " % rebalance_msg + str(
+                                              processor.node.account))
+
+                # verify processing
```

Comment:
Seems we are verifying processing twice: once here inside `maybe_upgrade_rolling_bounce_and_verify`, and once after `maybe_upgrade_rolling_bounce_and_verify` in `verify_processing`. Here we did the verify right after bouncing each node, and outside the `maybe_upgrade_rolling_bounce_and_verify` we just loop over them again and verify again. Is that intentional and necessary?

Reply:
Yes, it's intentional.  The check inside the `maybe_upgrade_rolling_bounce_and_verify` is for validating the individual node resumes processing after it gets bounced.  The additional check outside `maybe_upgrade_rolling_bounce_and_verify` confirms that all nodes are continuing to process and that rebalancing has ceased.  I can remove the outside check if you insist.

Reply:
That makes sense, let's keep it in that sense.

Reply:
The external check uses a new `monitor` object which when it's created sets an index into the current number of bytes in the file.  The `monitor` only checks in the file from that point forward.  The issue you describe could happen if you were to re-use an existing monitor.

Reply:
Fancy 👍